### PR TITLE
Support custom reason in @Disabled* / @Enabled* variants

### LINKED ISSUE #1787: Support custom reason in @Disabled* / @Enabled* variants
I would like to be able to write the reason why a test is disabled or enabled on a particular OS, JRE, etc. Currently the text is fixed in the different `ExecutionCondition`(s). 

I would propose that a new `default String reason() ""` is added to the variants and that reason is suffixed to the current default text for each condition.

## Deliverables

TBD

### PR DESCRIPTION
**Pull request by juliette-derancourt**
## Overview

Add the possibility to give a custom reason in case a test/container is disabled with one of the `@Enabled*`/`@Disabled*` annotations. This reason will be printed in addition to the current default reason.

Resolves #1787.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/BooleanExecutionCondition.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	@Override
+	public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
+		Optional<A> optional = findAnnotation(context.getElement(), annotationType);
+		if (optional.isPresent()) {
+			A annotation = optional.get();
+			String customReason = customDisabledReason.apply(annotation);
+			return isEnabled(annotation) ? enabled(enabledReason) : disabled(disabledReason, customReason);
+		}
+		return enabledByDefault();
```
[the remark this answers is not in the document]
Comment (juliette-derancourt) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/BooleanExecutionCondition.java:50]: You're right, it does look nicer 👍

**Comment by juliette-derancourt** [comment]
Thanks for the review @aepfli! 😄 
I also think it will be a useful feature, thanks @filiphr for suggesting this 😉

**Comment by filiphr** [comment]
Thanks a lot for implementing this @juliette-derancourt. Had a brief look and it looks even better then what I head in my head when I was suggesting it 😄.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ConditionEvaluationResult.java [hunk 2]
```diff
@@ -45,6 +45,21 @@ public static ConditionEvaluationResult disabled(String reason) {
...
+	/**
+	 * Factory for creating <em>disabled</em> results with custom reasons
+	 * added by the user.
+	 *
+	 * @param reason the default reason why the container or test should be disabled
+	 * @param customReason the custom reason why the container or test should be disabled
+	 * @return a disabled {@code ConditionEvaluationResult} with the given reasons
+	 */
+	public static ConditionEvaluationResult disabled(String reason, String customReason) {
+		if (customReason.isEmpty()) {
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ConditionEvaluationResult.java:57]: :o: Should we use `StringUtils.isBlank()` here?

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.7.0-M2.adoc [hunk 3]
```diff
@@ -39,7 +39,9 @@ on GitHub.
 
 ==== New Features and Improvements
 
-* ❓
+* All `@EnabledIf*`/`@DisabledIf*` annotations now have an optional `disabledReason`
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.7.0-M2.adoc:42]:
```suggestion
* All `@Enabled*`/`@Disabled*` annotations now have an optional `disabledReason`
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ConditionEvaluationResult.java (hunk 2 above)
Comment (juliette-derancourt) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ConditionEvaluationResult.java:57]: Good idea 👍

**Comment by juliette-derancourt** [comment]
> I'm wondering whether we should also mention it in the user guide but I don't think it's a must-have.

I think it would definitely be worth mentioning, but I didn't find an obvious place... Should it be at the top of the `Disabling Tests` section? But then it would be weird to mention it before the actual annotations... Or it could be in a note inside the first subsection? At the end of the section? I don't know 🤷‍♀️

**Comment by marcphilipp** [comment]
> > I'm wondering whether we should also mention it in the user guide but I don't think it's a must-have.
> 
> I think it would definitely be worth mentioning, but I didn't find an obvious place... Should it be at the top of the `Disabling Tests` section? But then it would be weird to mention it before the actual annotations... Or it could be in a note inside the first subsection? At the end of the section? I don't know 🤷‍♀️

How about in the first paragraph of [Conditional Test Execution](https://junit.org/junit5/docs/current/user-guide/#writing-tests-conditional-execution) where the built-in conditions are introduced?

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 4]
```diff
@@ -342,7 +342,9 @@ conditions _programmatically_. The simplest example of such a condition is the b
 several other annotation-based conditions in the `org.junit.jupiter.api.condition`
 package that allow developers to enable or disable containers and tests _declaratively_.
 When multiple `ExecutionCondition` extensions are registered, a container or test is
-disabled as soon as one of the conditions returns _disabled_.
+disabled as soon as one of the conditions returns _disabled_. If you wish to provide
+details about why they might be disabled, every annotation associated to these built-in
+conditions have a `disabledReason` attribute available for that purpose.
```
Comment (juliette-derancourt) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:347]: @marcphilipp something like that?
Should we also mention the fact that there is a default reason?

Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:347]:
```suggestion
disabled as soon as one of the conditions returns _disabled_. If you wish to provide
details about why they might be disabled, every annotation associated with these built-in
conditions has a `disabledReason` attribute available for that purpose.
```

Comment (marcphilipp) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:347]:
```suggestion
disabled as soon as one of the conditions returns _disabled_. If you wish to provide
details about why they might be disabled, every annotation associated with these built-in
conditions has a `disabledReason` attribute available for that purpose.
```

Comment (juliette-derancourt) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Indeed!
@sbrannen the fate of this sentence is now in your hands... 😁

**Pull request merged**